apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLog.java

```diff
@@ -0,0 +1,333 @@
+    }
+
+    @Override
+    public void debug(String s, Object o, Object o1) {
+        logger.debug(s, o, o1);
+    }
+
+    @Override
+    public void debug(String s, Object... objects) {
+        logger.debug(s, objects);
+        DeferredLogCache.debug(logger, s, objects);
+    }
+
+    @Override
+    public void debug(String s, Throwable throwable) {
+        logger.debug(s, throwable);
+    }
+
+    @Override
+    public boolean isDebugEnabled(Marker marker) {
+        return logger.isDebugEnabled(marker);
+    }
+
+    @Override
+    public void debug(Marker marker, String s) {
+        logger.debug(marker, s);
+    }
+
+    @Override
+    public void debug(Marker marker, String s, Object o) {
+        logger.debug(marker, s, o);
+    }
+
+    @Override
+    public void debug(Marker marker, String s, Object o, Object o1) {
+        logger.debug(marker, s, o, o1);
+    }
+
+    @Override
+    public void debug(Marker marker, String s, Object... objects) {
+        logger.debug(marker, s, objects);
+    }
+
+    @Override
+    public void debug(Marker marker, String s, Throwable throwable) {
+        logger.debug(marker, s, throwable);
+    }
+
+    @Override
+    public boolean isInfoEnabled() {
+        return logger.isInfoEnabled();
+    }
+
+    @Override
+    public void info(String s) {
+        logger.info(s);
+    }
+
+    @Override
+    public void info(String s, Object o) {
+        logger.info(s, o);
+    }
+
+    @Override
+    public void info(String s, Object o, Object o1) {
+        logger.info(s, o, o1);
+    }
+
+    @Override
+    public void info(String s, Object... objects) {
+        logger.info(s, objects);
+        DeferredLogCache.info(logger, s, objects);
+    }
+
+    @Override
+    public void info(String s, Throwable throwable) {
+        logger.info(s, throwable);
+    }
+
+    @Override
+    public boolean isInfoEnabled(Marker marker) {
+        return logger.isInfoEnabled(marker);
+    }
+
+    @Override
+    public void info(Marker marker, String s) {
+        logger.info(marker, s);
+    }
+
+    @Override
+    public void info(Marker marker, String s, Object o) {
+        logger.info(marker, s, o);
+    }
+
+    @Override
+    public void info(Marker marker, String s, Object o, Object o1) {
+        logger.info(marker, s, o, o1);
+    }
+
+    @Override
+    public void info(Marker marker, String s, Object... objects) {
+        logger.info(marker, s, objects);
+    }
+
+    @Override
+    public void info(Marker marker, String s, Throwable throwable) {
+        logger.info(marker, s, throwable);
+    }
+
+    @Override
+    public boolean isWarnEnabled() {
+        return logger.isWarnEnabled();
+    }
+
+    @Override
+    public void warn(String s) {
+        logger.warn(s);
+    }
+
+    @Override
+    public void warn(String s, Object o) {
+        logger.warn(s, o);
+    }
+
+    @Override
+    public void warn(String s, Object... objects) {
+        logger.warn(s, objects);
```

Comment:
I was thinking we add a check here so that we don't print the log twice? e.g.

```java
if (deferred) {
    DeferredLogCache.warn(logger, s, objects);
} else {
    logger.warn(s, objects);
}
```